miner/src/pool/listener.rs

```diff
@@ -116,6 +116,70 @@ impl txpool::Listener<Transaction> for Logger {
 	}
 }
 
+/// Transactions pool notifier
+#[derive(Default)]
+pub struct TransactionsPoolNotifier {
+	listeners: Vec<Arc<Mutex<mpsc::Sender<(H256, String)>>>>,
```

Comment:
@tomusdrw If I delete the `Arc<Mutex<`, it leads to errors

```
   Compiling ethcore v1.12.0 (/home/user/Work/parity-ethereum/ethcore)
error[E0277]: `std::sync::mpsc::Sender<ethcore_miner::pool::listener::TxStatus>` cannot be shared between threads safely
    --> ethcore/src/client/client.rs:1305:6
     |
1305 | impl snapshot::DatabaseRestore for Client {
     |      ^^^^^^^^^^^^^^^^^^^^^^^^^ `std::sync::mpsc::Sender<ethcore_miner::pool::listener::TxStatus>` cannot be shared between threads safely
     |
     = help: the trait `std::marker::Sync` is not implemented for `std::sync::mpsc::Sender<ethcore_miner::pool::listener::TxStatus>`
     = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<std::sync::mpsc::Sender<ethcore_miner::pool::listener::TxStatus>>`
     = note: required because it appears within the type `alloc::raw_vec::RawVec<std::sync::mpsc::Sender<ethcore_miner::pool::listener::TxStatus>>`
     = note: required because it appears within the type `std::vec::Vec<std::sync::mpsc::Sender<ethcore_miner::pool::listener::TxStatus>>`
     = note: required because it appears within the type `ethcore_miner::pool::listener::TransactionsPoolNotifier`
     = note: required because it appears within the type `(ethcore_miner::pool::listener::Logger, ethcore_miner::pool::listener::TransactionsPoolNotifier)`
     = note: required because it appears within the type `(ethcore_miner::pool::listener::Notifier, (ethcore_miner::pool::listener::Logger, ethcore_miner::pool::listener::TransactionsPoolNotifier))`
     = note: required because it appears within the type `(ethcore_miner::pool::local_transactions::LocalTransactionsList, (ethcore_miner::pool::listener::Notifier, (ethcore_miner::pool::listener::Logger, ethcore_miner::pool::listener::TransactionsPoolNotifier)))`
     = note: required because it appears within the type `transaction_pool::pool::Pool<ethcore_miner::pool::VerifiedTransaction, ethcore_miner::pool::scoring::NonceAndGasPrice, (ethcore_miner::pool::local_transactions::LocalTransactionsList, (ethcore_miner::pool::listener::Notifier, (ethcore_miner::pool::listener::Logger, ethcore_miner::pool::listener::TransactionsPoolNotifier)))>`
     = note: required because of the requirements on the impl of `std::marker::Sync` for `lock_api::rwlock::RwLock<parking_lot::RawRwLock, transaction_pool::pool::Pool<ethcore_miner::pool::VerifiedTransaction, ethcore_miner::pool::scoring::NonceAndGasPrice, (ethcore_miner::pool::local_transactions::LocalTransactionsList, (ethcore_miner::pool::listener::Notifier, (ethcore_miner::pool::listener::Logger, ethcore_miner::pool::listener::TransactionsPoolNotifier)))>>`
     = note: required because it appears within the type `ethcore_miner::pool::TransactionQueue`
     = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<ethcore_miner::pool::TransactionQueue>`
     = note: required because it appears within the type `miner::miner::Miner`
     = note: required because of the requirements on the impl of `std::marker::Sync` for `std::sync::Arc<miner::miner::Miner>`
     = note: required because it appears within the type `client::client::Importer`
     = note: required because it appears within the type `client::client::Client`
...
```
Found this: [Allowing Access from Multiple Threads with Sync](https://doc.rust-lang.org/stable/book/ch16-04-extensible-concurrency-sync-and-send.html#allowing-access-from-multiple-threads-with-sync)
And it seems to me that either Mutex is needed here, or I don't understand what am I doing wrong.

Reply:
Ah, I see. Yes, unfortunatelly we can't satisfy `Sync` with `std::mpsc::Sender`. To overcome this I'd go with `mpsc::unbounded_channel` from `futures` crate, this will also allow us to get rid of this extra thread responsible for polling the channel.